### PR TITLE
Fixes excessive exceptions from being thrown in case of graceful chan…

### PR DESCRIPTION
**Pull request by nayato**
…nel closure.

Motivation:
SocketException and ODE are being thrown when disconnecting gracefully.

Modifications:
- dismiss async read completion if it is done on inactive channel
- do not attempt sync-reading from a socket that is not Connected
- extra: fixing echo sample with better practices

Result:
No excessive errors happen during graceful shutdown

### CODE REVIEW

**Comment by nayato** [comment]
@mtuchkov, PTAL

**Comment by nayato** [comment]
Fixes #133

File: src/DotNetty.Transport/Channels/Sockets/AbstractSocketChannel.cs [hunk 1]
```diff
@@ -108,7 +108,18 @@ protected internal void ClearReadPending()
 
         protected void SetState(StateFlags stateToSet) => this.state |= stateToSet;
 
-        protected bool ResetState(StateFlags stateToReset)
+        /// <returns>state before modification</returns>
+        protected StateFlags ResetState(StateFlags stateToReset)
+        {
+            StateFlags oldState = this.state;
+            if ((oldState & stateToReset) != 0)
```
Comment (mtuchkov) [inline src/DotNetty.Transport/Channels/Sockets/AbstractSocketChannel.cs]: Is it possible that the old state is A&B and a new state is B&C ?

File: src/DotNetty.Transport/Channels/Sockets/TcpServerSocketChannel.cs [hunk 2]
```diff
@@ -127,10 +127,13 @@ public TcpServerSocketChannelUnsafe(TcpServerSocketChannel channel)
 
             public override void FinishRead(SocketChannelAsyncOperation operation)
             {
-                Contract.Requires(this.channel.EventLoop.InEventLoop);
+                Contract.Assert(this.channel.EventLoop.InEventLoop);
 
                 TcpServerSocketChannel ch = this.Channel;
-                ch.ResetState(StateFlags.ReadScheduled);
+                if ((ch.ResetState(StateFlags.ReadScheduled) & StateFlags.Active) == 0)
```
Comment (mtuchkov) [inline src/DotNetty.Transport/Channels/Sockets/TcpServerSocketChannel.cs]: can we come here when we already are in the ReadScheduled state?

File: src/DotNetty.Transport/Channels/Sockets/TcpSocketChannel.cs [hunk 3]
```diff
@@ -164,12 +164,18 @@ void OnConnected()
 
         protected override void DoClose()
         {
-            base.DoClose();
-            if (this.ResetState(StateFlags.Open | StateFlags.Active))
+            try
             {
-                this.Socket.Shutdown(SocketShutdown.Both);
-                this.Socket.Dispose();
+                if (this.TryResetState(StateFlags.Open | StateFlags.Active))
```
Comment (mtuchkov) [inline src/DotNetty.Transport/Channels/Sockets/TcpSocketChannel.cs]: On DoClose we try to set the Open&Active state. Is this correct?

File: src/DotNetty.Transport/Channels/Sockets/TcpServerSocketChannel.cs (hunk 2 above)
Comment (nayato) [inline src/DotNetty.Transport/Channels/Sockets/TcpServerSocketChannel.cs]: 100% ch is in ReadScheduled state.

---
In reply to: [73931101](https://github.com/Azure/DotNetty/pull/152#discussion_r73931101) [](ancestors = 73931101)

**Comment by mtuchkov** [comment]
:shipit:

File: src/DotNetty.Transport/Channels/Sockets/TcpServerSocketChannel.cs (hunk 2 above)
Comment (mtuchkov) [inline src/DotNetty.Transport/Channels/Sockets/TcpServerSocketChannel.cs]: OK

**Pull request merged**